clients/vscode-hlasmplugin/src/hlasmExternalFilesFtp.ts

```diff
@@ -47,171 +44,73 @@ class DatasetUriDetails implements ClientUriDetails {
-        try {
-            this.clientPool.closeClients();
-        }
-        finally {
-            this.clientSuspended = true;
-            this.stateChanged.fire(true);
-        }
-    }
-    resume() {
-        if (!this.clientSuspended) return;
-
-        this.clientSuspended = false;
-        this.stateChanged.fire(false);
-    }
-    suspended() {
-        return this.clientSuspended;
-    }
-
-    dispose(): void {
-        this.stateChanged.dispose();
-        this.clientPool.dispose();
-    }
-
-    private async getConnInfo() {
-        try {
-            if (this.clientSuspended)
-                throw new vscode.CancellationError();
-
-            if (this.activeConnectionInfo)
-                return this.activeConnectionInfo;
-
-            this.clientPool.closeClients();
-
-            const last = getLastRunConfig(this.context);
-            const connection = await gatherConnectionInfo(last);
-            await updateLastRunConfig(this.context, { host: connection.host, user: connection.user, jobcard: last.jobcard });
-
-            return connection;
-        }
-        catch (e) {
-            if (isCancellationError(e))
-                this.suspend();
-
-            throw e;
-        }
-    }
-
-    private async getConnectedClient() {
-        const client = new ftp.Client();
-        client.parseList = (rawList: string): ftp.FileInfo[] => {
-            return rawList.split(/\r?\n/).slice(1).filter(x => !/^\s*$/.test(x)).map(value => new ftp.FileInfo(value.split(/\s/, 1)[0]));
-        };
-        while (true) {
-            try {
-                await this.mutex.locked(async () => {
-                    const connection = await this.getConnInfo();
-                    this.activeConnectionInfo = null;
-                    await client.access(translateConnectionInfo(connection));
-                    this.activeConnectionInfo = connection;
-                });
-                return client;
-            }
-            catch (e) {
-                if (e instanceof ftp.FTPError) {
-                    vscode.window.showErrorMessage(e.message);
-                    continue;
-                }
-                this.clientPool.closeClients();
-                client.close();
-                throw e;
-            }
-        }
-    }
-
-    async listMembers(args: DatasetUriDetails): Promise<string[] | null> {
-        return this.clientPool.withClient(async (client) => {
-            try {
-                await checkedCommand(client, 'TYPE A');
-                checkResponse(await client.cd(`'${args.dataset}'`));
-                const list = await client.list();
-                return list.map(x => x.name);
-            }
-            catch (e) {
-                if (e instanceof ftp.FTPError && e.code == 550)
-                    return null;
-                throw e;
+export function HLASMExternalFilesFtp(context: vscode.ExtensionContext): ClientInterface<ConnectionInfo, DatasetUriDetails, DatasetUriDetails> {
```

Comment:
It is enough to use `ClientInterface<ConnectionInfo, ClientUriDetails, ClientUriDetails>` as the return type.

Reply:
That should actually not work, because the DatasetUriDetails is ClientUriDetails, but not all ClientUriDetails are DatasetUriDetails...